## Import Packages

In [2]:
import xarray as xr
import pandas as pd
import numpy as np
import holoviews as hv
from holoviews import opts
import dask.dataframe as dd
import hvplot.dask  # noqa

import numpy as np
import holoviews as hv
import pandas as pd
import numpy as np

import polars as pl

hv.extension('bokeh')


# Import Data

In [94]:
# 25 MB chunks 

df15 = dd.read_csv('ssp126_usa_onshore_daily_2015_2020.csv', blocksize=25e6, dtype={'lat': 'float64','lon': 'float64','sfcwind': 'float64','Capacity (MW)': 'float64', 'City' : 'string','Local area (taluk, county)' :'string','Major area (prefecture, district)':'string', 'State/Province':'string', 'Region':'string','Country':'string','sfc_wind80m':'float64','sfc_wind80m_floor':'float64'})
df21 = dd.read_csv('ssp126_usa_onshore_daily_2021_2030.csv', blocksize=25e6, dtype={'lat': 'float64','lon': 'float64','sfcwind': 'float64','Capacity (MW)': 'float64', 'City' : 'string','Local area (taluk, county)' :'string','Major area (prefecture, district)':'string', 'State/Province':'string', 'Region':'string','Country':'string','sfc_wind80m':'float64','sfc_wind80m_floor':'float64'})
df31 = dd.read_csv('ssp126_usa_onshore_daily_2031_2040.csv', blocksize=25e6, dtype={'lat': 'float64','lon': 'float64','sfcwind': 'float64','Capacity (MW)': 'float64', 'City' : 'string','Local area (taluk, county)' :'string','Major area (prefecture, district)':'string', 'State/Province':'string', 'Region':'string','Country':'string','sfc_wind80m':'float64','sfc_wind80m_floor':'float64'})
df41 = dd.read_csv('ssp126_usa_onshore_daily_2041_2050.csv', blocksize=25e6, dtype={'lat': 'float64','lon': 'float64','sfcwind': 'float64','Capacity (MW)': 'float64', 'City' : 'string','Local area (taluk, county)' :'string','Major area (prefecture, district)':'string', 'State/Province':'string', 'Region':'string','Country':'string','sfc_wind80m':'float64','sfc_wind80m_floor':'float64'})
df51 = dd.read_csv('ssp126_usa_onshore_daily_2051_2060.csv', blocksize=25e6, dtype={'lat': 'float64','lon': 'float64','sfcwind': 'float64','Capacity (MW)': 'float64', 'City' : 'string','Local area (taluk, county)' :'string','Major area (prefecture, district)':'string', 'State/Province':'string', 'Region':'string','Country':'string','sfc_wind80m':'float64','sfc_wind80m_floor':'float64'})
df61 = dd.read_csv('ssp126_usa_onshore_daily_2061_2070.csv', blocksize=25e6, dtype={'lat': 'float64','lon': 'float64','sfcwind': 'float64','Capacity (MW)': 'float64', 'City' : 'string','Local area (taluk, county)' :'string','Major area (prefecture, district)':'string', 'State/Province':'string', 'Region':'string','Country':'string','sfc_wind80m':'float64','sfc_wind80m_floor':'float64'})
df71 = dd.read_csv('ssp126_usa_onshore_daily_2071_2080.csv', blocksize=25e6, dtype={'lat': 'float64','lon': 'float64','sfcwind': 'float64','Capacity (MW)': 'float64', 'City' : 'string','Local area (taluk, county)' :'string','Major area (prefecture, district)':'string', 'State/Province':'string', 'Region':'string','Country':'string','sfc_wind80m':'float64','sfc_wind80m_floor':'float64'})
df81 = dd.read_csv('ssp126_usa_onshore_daily_2081_2090.csv', blocksize=25e6, dtype={'lat': 'float64','lon': 'float64','sfcwind': 'float64','Capacity (MW)': 'float64', 'City' : 'string','Local area (taluk, county)' :'string','Major area (prefecture, district)':'string', 'State/Province':'string', 'Region':'string','Country':'string','sfc_wind80m':'float64','sfc_wind80m_floor':'float64'})
df91 = dd.read_csv('ssp126_usa_onshore_daily_2091_2100.csv', blocksize=25e6, dtype={'lat': 'float64','lon': 'float64','sfcwind': 'float64','Capacity (MW)': 'float64', 'City' : 'string','Local area (taluk, county)' :'string','Major area (prefecture, district)':'string', 'State/Province':'string', 'Region':'string','Country':'string','sfc_wind80m':'float64','sfc_wind80m_floor':'float64'})

In [95]:
# merge dask dataframe to pandas dataframe
import dask.dataframe as dd
join = dd.concat([df15, df21])
join2 = dd.concat([join, df31])
join3 = dd.concat([join2, df41])
join4 = dd.concat([join3, df51])
join5 = dd.concat([join4, df61])
join6 = dd.concat([join5, df71])
join7 = dd.concat([join6, df81])
SSP126 = dd.concat([join7, df91])

# inspect results
# join.compute()

In [96]:
SSP126.tail()

,index,time,lat,lon,sfcwind,Capacity (MW),City,"Local area (taluk, county)","Major area (prefecture, district)",State/Province,Region,Country,sfc_wind80m,sfc_wind80m_floor
214646,23129,2100-12-27,42.75,-115.25,2.856540,21.0,<NA>,Twin Falls County,<NA>,Idaho,Americas,United States,3.845760,3.5
214647,23129,2100-12-28,42.75,-115.25,5.148798,21.0,<NA>,Twin Falls County,<NA>,Idaho,Americas,United States,6.931828,6.5
214648,23129,2100-12-29,42.75,-115.25,1.836774,21.0,<NA>,Twin Falls County,<NA>,Idaho,Americas,United States,2.472850,2.0
214649,23129,2100-12-30,42.75,-115.25,2.164511,21.0,<NA>,Twin Falls County,<NA>,Idaho,Americas,United States,2.914081,2.5
214650,23129,2100-12-31,42.75,-115.25,4.175478,21.0,<NA>,Twin Falls County,<NA>,Idaho,Americas,United States,5.621446,5.5


In [97]:
SSP126['Year'] = SSP126['time'].astype(str).str[:4] # creating year column
SSP126['Year'] = SSP126['Year'].astype(int) 
# add a column that is called 'loc' and it is the combination of lat and lon with a comma in between
SSP126['loc'] = SSP126['lat'].astype(str) + ',' + SSP126['lon'].astype(str)
SSP126.head()

,index,time,lat,lon,sfcwind,Capacity (MW),City,"Local area (taluk, county)","Major area (prefecture, district)",State/Province,Region,Country,sfc_wind80m,sfc_wind80m_floor,Year,loc
0,21508,2015-01-01,41.75,-93.75,3.373440,30.0,<NA>,Boone County,<NA>,Iowa,Americas,United States,4.541662,4.5,2015,"41.75,-93.75"
1,21508,2015-01-02,41.75,-93.75,3.341947,30.0,<NA>,Boone County,<NA>,Iowa,Americas,United States,4.499263,4.0,2015,"41.75,-93.75"
2,21508,2015-01-03,41.75,-93.75,3.919792,30.0,<NA>,Boone County,<NA>,Iowa,Americas,United States,5.277217,5.0,2015,"41.75,-93.75"
3,21508,2015-01-04,41.75,-93.75,2.496856,30.0,<NA>,Boone County,<NA>,Iowa,Americas,United States,3.361518,3.0,2015,"41.75,-93.75"
4,21508,2015-01-05,41.75,-93.75,4.605200,30.0,<NA>,Boone County,<NA>,Iowa,Americas,United States,6.199982,6.0,2015,"41.75,-93.75"


In [98]:
# split the 'SSP126' into datasets with the following years: 2015-2039, 2040-2064, 2065-2089, 2090-2100
SSP126_2015_2039 = SSP126[SSP126['Year'] <= 2039]
SSP126_2040_2064 = SSP126[(SSP126['Year'] >= 2040) & (SSP126['Year'] <= 2064)]
SSP126_2065_2089 = SSP126[(SSP126['Year'] >= 2065) & (SSP126['Year'] <= 2089)]
SSP126_2090_2100 = SSP126[SSP126['Year'] >= 2090]

Apply IEC Class

In [99]:
#group by loc and find the mean sfc_wind80m_floor. Uncomment and run one at a time. this takes a long time


SSP126_2015_2039_iec = SSP126_2015_2039.groupby('loc')['sfc_wind80m_floor'].mean().compute()
SSP126_2040_2064_iec = SSP126_2040_2064.groupby('loc')['sfc_wind80m_floor'].mean().compute()
SSP126_2065_2089_iec = SSP126_2065_2089.groupby('loc')['sfc_wind80m_floor'].mean().compute()
SSP126_2090_2100_iec = SSP126_2090_2100.groupby('loc')['sfc_wind80m_floor'].mean().compute()


In [100]:
SSP126_2015_2039_iec

loc
20.75,-156.25    7.700854
27.25,-97.25     5.931059
29.25,-100.25    5.021958
31.75,-100.75    5.962107
31.75,-101.75    6.029898
                   ...   
45.25,-119.75    4.032417
46.25,-118.25    3.576169
46.75,-120.25    2.878765
47.25,-120.25    2.744716
48.25,-100.75    5.380079
Name: sfc_wind80m_floor, Length: 444, dtype: float64

In [101]:
def IEC_Class(s): # s being sfc_wind80m_foor  25-year avg 
    if s < 7.5:
        return 1
    elif  7.5 <= s <= 8.5:
        return 2
    elif s > 8.5:
        return 3

In [102]:
#apply the IEC_Class function to the mean sfc_wind80m_floor in SSP126_2015_2039_iec and name the column 'iec'. Uncomment and run one at a time

SSP126_2015_2039_iec_n = SSP126_2015_2039_iec.apply(IEC_Class).rename('iec')
SSP126_2040_2064_iec_n = SSP126_2040_2064_iec.apply(IEC_Class).rename('iec')
SSP126_2065_2089_iec_n = SSP126_2065_2089_iec.apply(IEC_Class).rename('iec')
SSP126_2090_2100_iec_n = SSP126_2090_2100_iec.apply(IEC_Class).rename('iec')

In [103]:
# merge the two dataframes on loc. Uncomment and run one at a time
 
SSP126_2015_2039 = dd.merge(SSP126_2015_2039, SSP126_2015_2039_iec_n, on='loc', how='left')
SSP126_2040_2064 = dd.merge(SSP126_2040_2064, SSP126_2040_2064_iec_n, on='loc', how='left')
SSP126_2065_2089 = dd.merge(SSP126_2065_2089, SSP126_2065_2089_iec_n, on='loc', how='left')
SSP126_2090_2100 = dd.merge(SSP126_2090_2100, SSP126_2090_2100_iec_n, on='loc', how='left')

In [104]:
#rename 'sfc_wind80m_floor_x' to 'sfc_wind80m_floor'. Uncomment and run one at a time
SSP126_2015_2039 = SSP126_2015_2039.rename(columns={'sfc_wind80m_floor_x': 'sfc_wind80m_floor'})
SSP126_2040_2064 = SSP126_2040_2064.rename(columns={'sfc_wind80m_floor_x': 'sfc_wind80m_floor'})
SSP126_2065_2089 = SSP126_2065_2089.rename(columns={'sfc_wind80m_floor_x': 'sfc_wind80m_floor'})
SSP126_2090_2100 = SSP126_2090_2100.rename(columns={'sfc_wind80m_floor_x': 'sfc_wind80m_floor'})


In [105]:
IEC_Classifications = dd.read_csv('iec_classifications.csv')
# convert IEC_Class to i64
IEC_Classifications.compute()

,iec,CP_whole,sfc_wind80m_floor,Unnamed: 3,CP
0,1,19.3936,3.0,NaN,0.193936
1,1,33.1932,3.5,NaN,0.331932
2,1,37.9232,4.0,NaN,0.379232
3,1,41.4574,4.5,NaN,0.414574
4,1,42.1845,5.0,NaN,0.421845
...,...,...,...,...,...
130,3,1.8872,23.0,NaN,0.018872
131,3,1.6623,23.5,NaN,0.016623
132,3,1.4601,24.0,NaN,0.014601
133,3,1.2781,24.5,NaN,0.012781


Now we apply the Capacity Factors to each unique IEC Class

In [106]:
# merge the data on 'iec' and 'sfc_wind80m'. Uncomment and run one at a time
SSP126_2015_2039_n = dd.merge(SSP126_2015_2039, IEC_Classifications, on=['iec', 'sfc_wind80m_floor'], how='left')
SSP126_2040_2064_n = dd.merge(SSP126_2040_2064, IEC_Classifications, on=['iec', 'sfc_wind80m_floor'], how='left')
SSP126_2065_2089_n = dd.merge(SSP126_2065_2089, IEC_Classifications, on=['iec', 'sfc_wind80m_floor'], how='left')
SSP126_2090_2100_n = dd.merge(SSP126_2090_2100, IEC_Classifications, on=['iec', 'sfc_wind80m_floor'], how='left')

In [107]:
# average the CP by TEMELE_REGIONS_COPY and time. Uncomment and run one at a time. this takes a long time
SSP126_2015_2039_tt = SSP126_2015_2039_n.groupby(['State/Province', 'time'])['CP', 'Capacity (MW)'].mean().compute()
SSP126_2040_2064_tt = SSP126_2040_2064_n.groupby(['State/Province', 'time'])['CP', 'Capacity (MW)'].mean().compute()
SSP126_2065_2089_tt = SSP126_2065_2089_n.groupby(['State/Province', 'time'])['CP', 'Capacity (MW)'].mean().compute()
SSP126_2090_2100_tt = SSP126_2090_2100_n.groupby(['State/Province', 'time'])['CP', 'Capacity (MW)'].mean().compute()

In [108]:
SSP126_2015_2039_tt.head()

CP  Capacity (MW)
State/Province time                               
California     2015-01-01  0.193936      74.433735
               2015-01-02  0.203793      74.433735
               2015-01-03  0.193936      74.433735
               2015-01-04  0.193936      74.433735
               2015-01-05       NaN      74.433735

In [110]:
#save to csv. Uncomment and run one at a time

SSP126_2015_2039_tt.to_csv('SSP126_usa_onshore_2015_2039_test_state.csv')
SSP126_2040_2064_tt.to_csv('SSP126_usa_onshore_2040_2064_test_state.csv')
SSP126_2065_2089_tt.to_csv('SSP126_usa_onshore_2065_2089_test_state.csv')
SSP126_2090_2100_tt.to_csv('SSP126_usa_onshore_2090_2100_test_state.csv')

Organize in the Format that is compliant with TIMES/VEDA 2.0 --> Only Run the Block associated with the proper SSP

In [111]:
# dask open csv
SSP126_2015_2039_tt = dd.read_csv('SSP126_usa_onshore_2015_2039_test_state.csv')
SSP126_2040_2064_tt = dd.read_csv('SSP126_usa_onshore_2040_2064_test_state.csv')
SSP126_2065_2089_tt = dd.read_csv('SSP126_usa_onshore_2065_2089_test_state.csv')
SSP126_2090_2100_tt = dd.read_csv('SSP126_usa_onshore_2090_2100_test_state.csv')

# 2015-2039

In [112]:
#   SSP 126 2015-2039
#create a new column called 'Month' that is the month of the 'time' column
SSP126_2015_2039_tt['Month'] = SSP126_2015_2039_tt['time'].astype(str).str[5:7]
#create a new column called 'Day' that is the day of the 'time' column
SSP126_2015_2039_tt['Day'] = SSP126_2015_2039_tt['time'].astype(str).str[8:10]
#create a new column called 'TimeSlice' that is 'M' + Month + 'D' + Day
SSP126_2015_2039_tt['TimeSlice'] = 'M' + SSP126_2015_2039_tt['Month'] + 'D' + SSP126_2015_2039_tt['Day']
# add a new column called 'Attribute' that is entirely full of 'RNW'
SSP126_2015_2039_tt['Attribute'] = 'NCAP_AF'
# add a new column called 'Yr' that is the year of the 'time' column
SSP126_2015_2039_tt['Yr'] = SSP126_2015_2039_tt['time'].astype(str).str[:4]
# add a new column called 'PSET_CI' that is 'ELCWNO' for onshore or 'ELCWNF' for offshore
SSP126_2015_2039_tt['PSET_CI'] = 'ELCWNO'

SSP126_2015_2039_tt= SSP126_2015_2039_tt.groupby(['State/Province', 'time']).first().compute()
#pivot the table so that TEMELE_REGIONS_COPY are the columns, TimeSlice are the rows, and the values are the CP
SSP126_2015_2039_tt = SSP126_2015_2039_tt.pivot_table(index=['TimeSlice', 'Yr', 'Month', 'Day', 'time', 'Attribute', 'PSET_CI'], columns='State/Province', values='CP', aggfunc='first')
#replace all NaN values with 0
SSP126_2015_2039_tt = SSP126_2015_2039_tt.fillna(0)
# only keep TimeSlice', 'Yr', 'Month', 'Day', 'time', 'Attribute', 'PSET_CI', 'Virginia', 'Rhode Island'
#SSP126_2015_2039_tt = SSP126_2015_2039_tt[['Virginia', 'Rhode Island']]
SSP126_2015_2039_tt

State/Province                                           Alaska   Arizona  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2015 01    01  2015-01-01 NCAP_AF   ELCWNO   0.379232  0.193936   
          2016 01    01  2016-01-01 NCAP_AF   ELCWNO   0.286584  0.376895   
          2017 01    01  2017-01-01 NCAP_AF   ELCWNO   0.331932  0.331932   
          2018 01    01  2018-01-01 NCAP_AF   ELCWNO   0.379232  0.396230   
          2019 01    01  2019-01-01 NCAP_AF   ELCWNO   0.000000  0.397226   
...                                                         ...       ...   
M12D31    2035 12    31  2035-12-31 NCAP_AF   ELCWNO   0.000000  0.000000   
          2036 12    31  2036-12-31 NCAP_AF   ELCWNO   0.193936  0.418291   
          2037 12    31  2037-12-31 NCAP_AF   ELCWNO   0.379232  0.000000   
          2038 12    31  2038-12-31 NCAP_AF   ELCWNO   0.000000  0.323956   
          2039 12    31  2039-12-31 NCAP_AF   ELCWNO   0.193936  0.000000   

State/Province                                         California  Colorado  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                         
M01D01    2015 01    01  2015-01-01 NCAP_AF   ELCWNO     0.193936  0.323123   
          2016 01    01  2016-01-01 NCAP_AF   ELCWNO     0.219048  0.370889   
          2017 01    01  2017-01-01 NCAP_AF   ELCWNO     0.255046  0.438801   
          2018 01    01  2018-01-01 NCAP_AF   ELCWNO     0.193936  0.443643   
          2019 01    01  2019-01-01 NCAP_AF   ELCWNO     0.207171  0.432455   
...                                                           ...       ...   
M12D31    2035 12    31  2035-12-31 NCAP_AF   ELCWNO     0.331932  0.415058   
          2036 12    31  2036-12-31 NCAP_AF   ELCWNO     0.335349  0.440530   
          2037 12    31  2037-12-31 NCAP_AF   ELCWNO     0.000000  0.254597   
          2038 12    31  2038-12-31 NCAP_AF   ELCWNO     0.291212  0.422562   
          2039 12    31  2039-12-31 NCAP_AF   ELCWNO     0.000000  0.308085   

State/Province                                           Hawaii     Idaho  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2015 01    01  2015-01-01 NCAP_AF   ELCWNO   0.364480  0.262934   
          2016 01    01  2016-01-01 NCAP_AF   ELCWNO   0.312216  0.371316   
          2017 01    01  2017-01-01 NCAP_AF   ELCWNO   0.460097  0.308381   
          2018 01    01  2018-01-01 NCAP_AF   ELCWNO   0.364692  0.378624   
          2019 01    01  2019-01-01 NCAP_AF   ELCWNO   0.390575  0.380502   
...                                                         ...       ...   
M12D31    2035 12    31  2035-12-31 NCAP_AF   ELCWNO   0.425731  0.247601   
          2036 12    31  2036-12-31 NCAP_AF   ELCWNO   0.376312  0.000000   
          2037 12    31  2037-12-31 NCAP_AF   ELCWNO   0.366463  0.262934   
          2038 12    31  2038-12-31 NCAP_AF   ELCWNO   0.326631  0.193936   
          2039 12    31  2039-12-31 NCAP_AF   ELCWNO   0.403689  0.193936   

State/Province                                         Illinois   Indiana  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2015 01    01  2015-01-01 NCAP_AF   ELCWNO   0.408622  0.438507   
          2016 01    01  2016-01-01 NCAP_AF   ELCWNO   0.418824  0.407506   
          2017 01    01  2017-01-01 NCAP_AF   ELCWNO   0.259675  0.233165   
          2018 01    01  2018-01-01 NCAP_AF   ELCWNO   0.306546  0.379232   
          2019 01    01  2019-01-01 NCAP_AF   ELCWNO   0.437802  0.440924   
...                                                         ...       ...   
M12D31    2035 12    31  2035-12-31 NCAP_AF   ELCWNO   0.385002  0.394659   
          2036 12    31  2036-12-31 NCAP_AF   ELCWNO   0.371510  0.331932   
          2037 12    31  2037-12-31 NCAP_AF   ELCWNO   0.435377  0.421118   
          2038 12    31  2038-12-31 NCAP_AF   ELCWNO   0.347292  0.393498   
          2039 12    31  2039-12-31 NCAP_AF 

In [113]:
#export to csv
SSP126_2015_2039_tt.to_csv('SSP126_usa_onshore_2015_2039_TIMES_state.csv')

# 2040-2064

In [114]:
#create a new column called 'Month' that is the month of the 'time' column
SSP126_2040_2064_tt['Month'] = SSP126_2040_2064_tt['time'].astype(str).str[5:7]
#create a new column called 'Day' that is the day of the 'time' column
SSP126_2040_2064_tt['Day'] = SSP126_2040_2064_tt['time'].astype(str).str[8:10]
#create a new column called 'TimeSlice' that is 'M' + Month + 'D' + Day
SSP126_2040_2064_tt['TimeSlice'] = 'M' + SSP126_2040_2064_tt['Month'] + 'D' + SSP126_2040_2064_tt['Day']
# add a new column called 'Attribute' that is entirely full of 'RNW'
SSP126_2040_2064_tt['Attribute'] = 'NCAP_AF'
# add a new column called 'Yr' that is the year of the 'time' column
SSP126_2040_2064_tt['Yr'] = SSP126_2040_2064_tt['time'].astype(str).str[:4]
# add a new column called 'PSET_CI' that is 'ELCWNO' for onshore or 'ELCWNF' for offshore
SSP126_2040_2064_tt['PSET_CI'] = 'ELCWNO'


SSP126_2040_2064_tt= SSP126_2040_2064_tt.groupby(['State/Province', 'time']).first().compute()
#pivot the table so that TEMELE_REGIONS_COPY are the columns, TimeSlice are the rows, and the values are the CP
SSP126_2040_2064_tt = SSP126_2040_2064_tt.pivot_table(index=['TimeSlice', 'Yr', 'Month', 'Day', 'time', 'Attribute', 'PSET_CI'], columns='State/Province', values='CP', aggfunc='first')
#replace all NaN values with 0
SSP126_2040_2064_tt = SSP126_2040_2064_tt.fillna(0)
# only keep TimeSlice', 'Yr', 'Month', 'Day', 'time', 'Attribute', 'PSET_CI', 'Virginia', 'Rhode Island'
#SSP126_2040_2064_tt = SSP126_2040_2064_tt[['Virginia', 'Rhode Island']]
SSP126_2040_2064_tt

State/Province                                           Alaska   Arizona  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2040 01    01  2040-01-01 NCAP_AF   ELCWNO   0.315406  0.342390   
          2041 01    01  2041-01-01 NCAP_AF   ELCWNO   0.000000  0.000000   
          2042 01    01  2042-01-01 NCAP_AF   ELCWNO   0.000000  0.367407   
          2043 01    01  2043-01-01 NCAP_AF   ELCWNO   0.193936  0.366578   
          2044 01    01  2044-01-01 NCAP_AF   ELCWNO   0.193936  0.379232   
...                                                         ...       ...   
M12D31    2060 12    31  2060-12-31 NCAP_AF   ELCWNO   0.000000  0.337390   
          2061 12    31  2061-12-31 NCAP_AF   ELCWNO   0.315406  0.328988   
          2062 12    31  2062-12-31 NCAP_AF   ELCWNO   0.193936  0.000000   
          2063 12    31  2063-12-31 NCAP_AF   ELCWNO   0.379232  0.337390   
          2064 12    31  2064-12-31 NCAP_AF   ELCWNO   0.000000  0.000000   

State/Province                                         California  Colorado  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                         
M01D01    2040 01    01  2040-01-01 NCAP_AF   ELCWNO     0.193936  0.349847   
          2041 01    01  2041-01-01 NCAP_AF   ELCWNO     0.193936  0.433664   
          2042 01    01  2042-01-01 NCAP_AF   ELCWNO     0.322068  0.448477   
          2043 01    01  2043-01-01 NCAP_AF   ELCWNO     0.193936  0.429748   
          2044 01    01  2044-01-01 NCAP_AF   ELCWNO     0.193936  0.419945   
...                                                           ...       ...   
M12D31    2060 12    31  2060-12-31 NCAP_AF   ELCWNO     0.297433  0.407135   
          2061 12    31  2061-12-31 NCAP_AF   ELCWNO     0.400446  0.443265   
          2062 12    31  2062-12-31 NCAP_AF   ELCWNO     0.383872  0.408077   
          2063 12    31  2063-12-31 NCAP_AF   ELCWNO     0.379232  0.228435   
          2064 12    31  2064-12-31 NCAP_AF   ELCWNO     0.331932  0.421035   

State/Province                                           Hawaii     Idaho  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2040 01    01  2040-01-01 NCAP_AF   ELCWNO   0.444821  0.000000   
          2041 01    01  2041-01-01 NCAP_AF   ELCWNO   0.349082  0.193936   
          2042 01    01  2042-01-01 NCAP_AF   ELCWNO   0.400913  0.193936   
          2043 01    01  2043-01-01 NCAP_AF   ELCWNO   0.299454  0.367848   
          2044 01    01  2044-01-01 NCAP_AF   ELCWNO   0.331621  0.000000   
...                                                         ...       ...   
M12D31    2060 12    31  2060-12-31 NCAP_AF   ELCWNO   0.427227  0.423700   
          2061 12    31  2061-12-31 NCAP_AF   ELCWNO   0.401768  0.337059   
          2062 12    31  2062-12-31 NCAP_AF   ELCWNO   0.380718  0.262934   
          2063 12    31  2063-12-31 NCAP_AF   ELCWNO   0.370302  0.372967   
          2064 12    31  2064-12-31 NCAP_AF   ELCWNO   0.308544  0.436614   

State/Province                                         Illinois   Indiana  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2040 01    01  2040-01-01 NCAP_AF   ELCWNO   0.245283  0.304333   
          2041 01    01  2041-01-01 NCAP_AF   ELCWNO   0.424458  0.421845   
          2042 01    01  2042-01-01 NCAP_AF   ELCWNO   0.452333  0.455580   
          2043 01    01  2043-01-01 NCAP_AF   ELCWNO   0.375623  0.421118   
          2044 01    01  2044-01-01 NCAP_AF   ELCWNO   0.409500  0.437862   
...                                                         ...       ...   
M12D31    2060 12    31  2060-12-31 NCAP_AF   ELCWNO   0.447269  0.450056   
          2061 12    31  2061-12-31 NCAP_AF   ELCWNO   0.398536  0.304333   
          2062 12    31  2062-12-31 NCAP_AF   ELCWNO   0.453615  0.455395   
          2063 12    31  2063-12-31 NCAP_AF   ELCWNO   0.417445  0.409273   
          2064 12    31  2064-12-31 NCAP_AF 

In [115]:
#export to csv
SSP126_2040_2064_tt.to_csv('SSP126_usa_onshore_2040_2064_TIMES_state.csv')

# 2065-2089

In [116]:
#   SSP 126 2065-2089
#create a new column called 'Month' that is the month of the 'time' column
SSP126_2065_2089_tt['Month'] = SSP126_2065_2089_tt['time'].astype(str).str[5:7]
#create a new column called 'Day' that is the day of the 'time' column
SSP126_2065_2089_tt['Day'] = SSP126_2065_2089_tt['time'].astype(str).str[8:10]
#create a new column called 'TimeSlice' that is 'M' + Month + 'D' + Day
SSP126_2065_2089_tt['TimeSlice'] = 'M' + SSP126_2065_2089_tt['Month'] + 'D' + SSP126_2065_2089_tt['Day']
# add a new column called 'Attribute' that is entirely full of 'RNW'
SSP126_2065_2089_tt['Attribute'] = 'NCAP_AF'
# add a new column called 'Yr' that is the year of the 'time' column
SSP126_2065_2089_tt['Yr'] = SSP126_2065_2089_tt['time'].astype(str).str[:4]
# add a new column called 'PSET_CI' that is 'ELCWNO' for onshore or 'ELCWNF' for offshore
SSP126_2065_2089_tt['PSET_CI'] = 'ELCWNO'

SSP126_2065_2089_tt= SSP126_2065_2089_tt.groupby(['State/Province', 'time']).first().compute()
#pivot the table so that TEMELE_REGIONS_COPY are the columns, TimeSlice are the rows, and the values are the CP
SSP126_2065_2089_tt = SSP126_2065_2089_tt.pivot_table(index=['TimeSlice', 'Yr', 'Month', 'Day', 'time', 'Attribute', 'PSET_CI'], columns='State/Province', values='CP', aggfunc='first')
#replace all NaN values with 0
SSP126_2065_2089_tt = SSP126_2065_2089_tt.fillna(0)
# only keep TimeSlice', 'Yr', 'Month', 'Day', 'time', 'Attribute', 'PSET_CI', 'Virginia', 'Rhode Island'
#SSP126_2065_2089_tt = SSP126_2065_2089_tt[['Virginia', 'Rhode Island']]
SSP126_2065_2089_tt

State/Province                                           Alaska   Arizona  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2065 01    01  2065-01-01 NCAP_AF   ELCWNO   0.331932  0.363378   
          2066 01    01  2066-01-01 NCAP_AF   ELCWNO   0.379232  0.193936   
          2067 01    01  2067-01-01 NCAP_AF   ELCWNO   0.000000  0.418542   
          2068 01    01  2068-01-01 NCAP_AF   ELCWNO   0.331932  0.330590   
          2069 01    01  2069-01-01 NCAP_AF   ELCWNO   0.000000  0.193936   
...                                                         ...       ...   
M12D31    2085 12    31  2085-12-31 NCAP_AF   ELCWNO   0.331932  0.238064   
          2086 12    31  2086-12-31 NCAP_AF   ELCWNO   0.000000  0.268054   
          2087 12    31  2087-12-31 NCAP_AF   ELCWNO   0.193936  0.000000   
          2088 12    31  2088-12-31 NCAP_AF   ELCWNO   0.379232  0.359479   
          2089 12    31  2089-12-31 NCAP_AF   ELCWNO   0.373253  0.193936   

State/Province                                         California  Colorado  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                         
M01D01    2065 01    01  2065-01-01 NCAP_AF   ELCWNO     0.203793  0.429388   
          2066 01    01  2066-01-01 NCAP_AF   ELCWNO     0.000000  0.343002   
          2067 01    01  2067-01-01 NCAP_AF   ELCWNO     0.277416  0.417633   
          2068 01    01  2068-01-01 NCAP_AF   ELCWNO     0.273349  0.428635   
          2069 01    01  2069-01-01 NCAP_AF   ELCWNO     0.193936  0.350560   
...                                                           ...       ...   
M12D31    2085 12    31  2085-12-31 NCAP_AF   ELCWNO     0.329251  0.406019   
          2086 12    31  2086-12-31 NCAP_AF   ELCWNO     0.309853  0.335303   
          2087 12    31  2087-12-31 NCAP_AF   ELCWNO     0.203793  0.432167   
          2088 12    31  2088-12-31 NCAP_AF   ELCWNO     0.268306  0.278122   
          2089 12    31  2089-12-31 NCAP_AF   ELCWNO     0.228435  0.273418   

State/Province                                           Hawaii     Idaho  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2065 01    01  2065-01-01 NCAP_AF   ELCWNO   0.391295  0.265313   
          2066 01    01  2066-01-01 NCAP_AF   ELCWNO   0.263071  0.331932   
          2067 01    01  2067-01-01 NCAP_AF   ELCWNO   0.243912  0.393477   
          2068 01    01  2068-01-01 NCAP_AF   ELCWNO   0.355720  0.193936   
          2069 01    01  2069-01-01 NCAP_AF   ELCWNO   0.335145  0.419867   
...                                                         ...       ...   
M12D31    2085 12    31  2085-12-31 NCAP_AF   ELCWNO   0.452825  0.326319   
          2086 12    31  2086-12-31 NCAP_AF   ELCWNO   0.374249  0.379322   
          2087 12    31  2087-12-31 NCAP_AF   ELCWNO   0.353785  0.000000   
          2088 12    31  2088-12-31 NCAP_AF   ELCWNO   0.390744  0.397165   
          2089 12    31  2089-12-31 NCAP_AF   ELCWNO   0.434557  0.193936   

State/Province                                         Illinois   Indiana  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2065 01    01  2065-01-01 NCAP_AF   ELCWNO   0.454259  0.455634   
          2066 01    01  2066-01-01 NCAP_AF   ELCWNO   0.443759  0.454348   
          2067 01    01  2067-01-01 NCAP_AF   ELCWNO   0.423341  0.423761   
          2068 01    01  2068-01-01 NCAP_AF   ELCWNO   0.399183  0.417846   
          2069 01    01  2069-01-01 NCAP_AF   ELCWNO   0.442419  0.433958   
...                                                         ...       ...   
M12D31    2085 12    31  2085-12-31 NCAP_AF   ELCWNO   0.420371  0.424656   
          2086 12    31  2086-12-31 NCAP_AF   ELCWNO   0.275229  0.290533   
          2087 12    31  2087-12-31 NCAP_AF   ELCWNO   0.445725  0.449022   
          2088 12    31  2088-12-31 NCAP_AF   ELCWNO   0.394612  0.351242   
          2089 12    31  2089-12-31 NCAP_AF 

In [117]:
#export to csv
SSP126_2065_2089_tt.to_csv('SSP126_usa_onshore_2065_2089_TIMES_state.csv')


# 2090-2100

In [118]:
#   SSP 126 2090-2100

#create a new column called 'Month' that is the month of the 'time' column
SSP126_2090_2100_tt['Month'] = SSP126_2090_2100_tt['time'].astype(str).str[5:7]
#create a new column called 'Day' that is the day of the 'time' column
SSP126_2090_2100_tt['Day'] = SSP126_2090_2100_tt['time'].astype(str).str[8:10]
#create a new column called 'TimeSlice' that is 'M' + Month + 'D' + Day
SSP126_2090_2100_tt['TimeSlice'] = 'M' + SSP126_2090_2100_tt['Month'] + 'D' + SSP126_2090_2100_tt['Day']
# add a new column called 'Attribute' that is entirely full of 'RNW'
SSP126_2090_2100_tt['Attribute'] = 'NCAP_AF'
# add a new column called 'Yr' that is the year of the 'time' column
SSP126_2090_2100_tt['Yr'] = SSP126_2090_2100_tt['time'].astype(str).str[:4]
# add a new column called 'PSET_CI' that is 'ELCWNO' for onshore or 'ELCWNF' for offshore
SSP126_2090_2100_tt['PSET_CI'] = 'ELCWNO'


SSP126_2090_2100_tt= SSP126_2090_2100_tt.groupby(['State/Province', 'time']).first().compute()
#pivot the table so that TEMELE_REGIONS_COPY are the columns, TimeSlice are the rows, and the values are the CP
SSP126_2090_2100_tt = SSP126_2090_2100_tt.pivot_table(index=['TimeSlice', 'Yr', 'Month', 'Day', 'time', 'Attribute', 'PSET_CI'], columns='State/Province', values='CP', aggfunc='first')
#replace all NaN values with 0
SSP126_2090_2100_tt = SSP126_2090_2100_tt.fillna(0)

# only keep TimeSlice', 'Yr', 'Month', 'Day', 'time', 'Attribute', 'PSET_CI', 'Virginia', 'Rhode Island'
#SSP126_2090_2100_tt = SSP126_2090_2100_tt[['Virginia', 'Rhode Island']]


In [119]:
SSP126_2090_2100_tt

State/Province                                           Alaska   Arizona  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2090 01    01  2090-01-01 NCAP_AF   ELCWNO   0.373253  0.286584   
          2091 01    01  2091-01-01 NCAP_AF   ELCWNO   0.286584  0.000000   
          2092 01    01  2092-01-01 NCAP_AF   ELCWNO   0.000000  0.000000   
          2093 01    01  2093-01-01 NCAP_AF   ELCWNO   0.000000  0.409875   
          2094 01    01  2094-01-01 NCAP_AF   ELCWNO   0.286584  0.414376   
...                                                         ...       ...   
M12D31    2096 12    31  2096-12-31 NCAP_AF   ELCWNO   0.436877  0.000000   
          2097 12    31  2097-12-31 NCAP_AF   ELCWNO   0.436877  0.000000   
          2098 12    31  2098-12-31 NCAP_AF   ELCWNO   0.262934  0.000000   
          2099 12    31  2099-12-31 NCAP_AF   ELCWNO   0.286584  0.420715   
          2100 12    31  2100-12-31 NCAP_AF   ELCWNO   0.436877  0.000000   

State/Province                                         California  Colorado  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                         
M01D01    2090 01    01  2090-01-01 NCAP_AF   ELCWNO     0.193936  0.427342   
          2091 01    01  2091-01-01 NCAP_AF   ELCWNO     0.412891  0.430728   
          2092 01    01  2092-01-01 NCAP_AF   ELCWNO     0.204968  0.441314   
          2093 01    01  2093-01-01 NCAP_AF   ELCWNO     0.335874  0.273990   
          2094 01    01  2094-01-01 NCAP_AF   ELCWNO     0.364110  0.426356   
...                                                           ...       ...   
M12D31    2096 12    31  2096-12-31 NCAP_AF   ELCWNO     0.193936  0.361812   
          2097 12    31  2097-12-31 NCAP_AF   ELCWNO     0.193936  0.251434   
          2098 12    31  2098-12-31 NCAP_AF   ELCWNO     0.000000  0.368305   
          2099 12    31  2099-12-31 NCAP_AF   ELCWNO     0.343285  0.414903   
          2100 12    31  2100-12-31 NCAP_AF   ELCWNO     0.193936  0.421668   

State/Province                                           Hawaii     Idaho  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2090 01    01  2090-01-01 NCAP_AF   ELCWNO   0.387657  0.349724   
          2091 01    01  2091-01-01 NCAP_AF   ELCWNO   0.390321  0.331932   
          2092 01    01  2092-01-01 NCAP_AF   ELCWNO   0.429870  0.388481   
          2093 01    01  2093-01-01 NCAP_AF   ELCWNO   0.384936  0.193936   
          2094 01    01  2094-01-01 NCAP_AF   ELCWNO   0.377857  0.000000   
...                                                         ...       ...   
M12D31    2096 12    31  2096-12-31 NCAP_AF   ELCWNO   0.270562  0.322161   
          2097 12    31  2097-12-31 NCAP_AF   ELCWNO   0.388807  0.193936   
          2098 12    31  2098-12-31 NCAP_AF   ELCWNO   0.376164  0.267534   
          2099 12    31  2099-12-31 NCAP_AF   ELCWNO   0.406899  0.000000   
          2100 12    31  2100-12-31 NCAP_AF   ELCWNO   0.315110  0.397278   

State/Province                                         Illinois   Indiana  \
TimeSlice Yr   Month Day time       Attribute PSET_CI                       
M01D01    2090 01    01  2090-01-01 NCAP_AF   ELCWNO   0.322816  0.383935   
          2091 01    01  2091-01-01 NCAP_AF   ELCWNO   0.312915  0.416128   
          2092 01    01  2092-01-01 NCAP_AF   ELCWNO   0.414582  0.428609   
          2093 01    01  2093-01-01 NCAP_AF   ELCWNO   0.193936  0.193936   
          2094 01    01  2094-01-01 NCAP_AF   ELCWNO   0.434312  0.442886   
...                                                         ...       ...   
M12D31    2096 12    31  2096-12-31 NCAP_AF   ELCWNO   0.453190  0.447735   
          2097 12    31  2097-12-31 NCAP_AF   ELCWNO   0.393868  0.419351   
          2098 12    31  2098-12-31 NCAP_AF   ELCWNO   0.347733  0.422721   
          2099 12    31  2099-12-31 NCAP_AF   ELCWNO   0.446841  0.451885   
          2100 12    31  2100-12-31 NCAP_AF 

In [120]:
#export to csv
SSP126_2090_2100_tt.to_csv('SSP126_usa_onshore_2090_2100_TIMES_state.csv')

# Organize into TIMES Time Slices 

In [121]:
import polars as pl

# Read all the *TIMES.csv files in Polars and concatenate them
SSP126_TIMES = pl.read_csv('SSP126_usa_onshore_2015_2039_TIMES_state.csv')
SSP126_TIMES = SSP126_TIMES.extend(pl.read_csv('SSP126_usa_onshore_2040_2064_TIMES_state.csv'))
SSP126_TIMES = SSP126_TIMES.extend(pl.read_csv('SSP126_usa_onshore_2065_2089_TIMES_state.csv'))
SSP126_TIMES = SSP126_TIMES.extend(pl.read_csv('SSP126_usa_onshore_2090_2100_TIMES_state.csv'))

# Filter the DataFrame to keep only rows with 'Yr' greater than or equal to 2020
SSP126_TIMES = SSP126_TIMES.filter(pl.col('Yr') >= 2020)

In [122]:
SSP126_TIMES.head()

TimeSlice,Yr,Month,Day,time,Attribute,PSET_CI,Alaska,Arizona,California,Colorado,Hawaii,Idaho,Illinois,Indiana,Iowa,Kansas,Maine,Maryland,Massachusetts,Michigan,Minnesota,Missouri,Montana,Nebraska,Nevada,New Hampshire,New Mexico,New York,North Carolina,North Dakota,Ohio,Oklahoma,Oregon,Pennsylvania,South Dakota,Tennessee,Texas,Utah,Vermont,Washington,West Virginia,Wisconsin,Wyoming
str,i64,i64,i64,str,str,str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
"""M01D01""",2020,1,1,"""2020-01-01""","""NCAP_AF""","""ELCWNO""",0.193936,0.3354928,0.322932,0.382723,0.372024,0.378799,0.443297,0.448642,0.429492,0.383685,0.2680544,0.331932,0.331932,0.447255,0.307729,0.422332,0.347387,0.302281,0.0,0.262934,0.432635,0.434391,0.193936,0.2495248,0.448662,0.4425755,0.429215,0.343343,0.234426,0.421845,0.442901,0.0,0.355582,0.411531,0.331932,0.425844,0.401381
"""M01D01""",2021,1,1,"""2021-01-01""","""NCAP_AF""","""ELCWNO""",0.262934,0.255701,0.209329,0.413452,0.437071,0.193936,0.427397,0.4361254,0.436697,0.39941,0.447172,0.2309952,0.421845,0.443874,0.445556,0.400819,0.395742,0.435994,0.0,0.437213,0.355668,0.383725,0.193936,0.451936,0.436877,0.353921,0.0,0.356882,0.454623,0.0,0.368371,0.193936,0.421845,0.193936,0.193936,0.4295778,0.429521
"""M01D01""",2022,1,1,"""2022-01-01""","""NCAP_AF""","""ELCWNO""",0.448366,0.419421,0.421845,0.421046,0.388375,0.432419,0.245169,0.425883,0.193936,0.389514,0.286367,0.3863004,0.0,0.219026,0.249722,0.414574,0.325591,0.228567,0.0,0.0,0.430069,0.283545,0.414574,0.384788,0.283446,0.372789,0.371596,0.334196,0.289875,0.421845,0.429984,0.414574,0.193936,0.373321,0.398171,0.0,0.446337
"""M01D01""",2023,1,1,"""2023-01-01""","""NCAP_AF""","""ELCWNO""",0.0,0.193936,0.419816,0.333917,0.361348,0.371508,0.388343,0.4142611,0.21863,0.273358,0.352738,0.2309952,0.379232,0.441279,0.299279,0.193936,0.4166806,0.313134,0.0,0.3232528,0.441934,0.430678,0.0,0.431892,0.42614,0.371262,0.415731,0.350579,0.393496,0.331932,0.409575,0.0,0.355582,0.413848,0.224602,0.379906,0.40699
"""M01D01""",2024,1,1,"""2024-01-01""","""NCAP_AF""","""ELCWNO""",0.193936,0.262934,0.193936,0.443864,0.359826,0.41041,0.43845,0.4064656,0.444845,0.43347,0.338858,0.2309952,0.379232,0.355065,0.45181,0.431095,0.406374,0.4370968,0.0,0.297433,0.415498,0.429774,0.45518,0.283774,0.396456,0.427897,0.382378,0.305921,0.426041,0.193936,0.437992,0.193936,0.3762425,0.392115,0.265779,0.4342062,0.242139


In [123]:
import polars as pl

# Add 'Season' column based on the 'Month' column
SSP126_TIMES = SSP126_TIMES.with_columns(
    Season=
    SSP126_TIMES["Month"].apply(
        lambda month: "S01" if month in [1, 2, 12] else "S02" if month in [3, 4, 5] else "S03" if month in [6, 7, 8] else "S04"
    ),
)

#read 'time' column as datetime 

SSP126_TIMES = SSP126_TIMES.with_columns(pl.col('time').str.strptime(pl.Datetime).alias('time'))

# find the weekday number of the 'time' column 
SSP126_TIMES = SSP126_TIMES.with_columns(
    Weekday=
    SSP126_TIMES["time"].apply(
        lambda time: time.isoweekday()
    ),
)

# if the 'Weekday' is equal to 6, 2, 3, 4, or 5, then the 'Type' is 'WD2', otherwise it is 'WD1'
SSP126_TIMES = SSP126_TIMES.with_columns(
    Type=
    SSP126_TIMES["Weekday"].apply(
        lambda Weekday: "WD2" if Weekday in [6, 2, 3, 4, 5] else "WD1"
    ),
)

SSP126_TIMES = SSP126_TIMES.with_columns(pl.lit('NCAP_AF').alias('Attribute'))


In [124]:
SSP126_TIMES

TimeSlice,Yr,Month,Day,time,Attribute,PSET_CI,Alaska,Arizona,California,Colorado,Hawaii,Idaho,Illinois,Indiana,Iowa,Kansas,Maine,Maryland,Massachusetts,Michigan,Minnesota,Missouri,Montana,Nebraska,Nevada,New Hampshire,New Mexico,New York,North Carolina,North Dakota,Ohio,Oklahoma,Oregon,Pennsylvania,South Dakota,Tennessee,Texas,Utah,Vermont,Washington,West Virginia,Wisconsin,Wyoming,Season,Weekday,Type
str,i64,i64,i64,datetime[μs],str,str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,str,i64,str
"""M01D01""",2020,1,1,2020-01-01 00:00:00,"""NCAP_AF""","""ELCWNO""",0.193936,0.3354928,0.322932,0.382723,0.372024,0.378799,0.443297,0.448642,0.429492,0.383685,0.2680544,0.331932,0.331932,0.447255,0.307729,0.422332,0.347387,0.302281,0.0,0.262934,0.432635,0.434391,0.193936,0.2495248,0.448662,0.4425755,0.429215,0.343343,0.234426,0.421845,0.442901,0.0,0.355582,0.411531,0.331932,0.425844,0.401381,"""S01""",3,"""WD2"""
"""M01D01""",2021,1,1,2021-01-01 00:00:00,"""NCAP_AF""","""ELCWNO""",0.262934,0.255701,0.209329,0.413452,0.437071,0.193936,0.427397,0.4361254,0.436697,0.39941,0.447172,0.2309952,0.421845,0.443874,0.445556,0.400819,0.395742,0.435994,0.0,0.437213,0.355668,0.383725,0.193936,0.451936,0.436877,0.353921,0.0,0.356882,0.454623,0.0,0.368371,0.193936,0.421845,0.193936,0.193936,0.4295778,0.429521,"""S01""",5,"""WD2"""
"""M01D01""",2022,1,1,2022-01-01 00:00:00,"""NCAP_AF""","""ELCWNO""",0.448366,0.419421,0.421845,0.421046,0.388375,0.432419,0.245169,0.425883,0.193936,0.389514,0.286367,0.3863004,0.0,0.219026,0.249722,0.414574,0.325591,0.228567,0.0,0.0,0.430069,0.283545,0.414574,0.384788,0.283446,0.372789,0.371596,0.334196,0.289875,0.421845,0.429984,0.414574,0.193936,0.373321,0.398171,0.0,0.446337,"""S01""",6,"""WD2"""
"""M01D01""",2023,1,1,2023-01-01 00:00:00,"""NCAP_AF""","""ELCWNO""",0.0,0.193936,0.419816,0.333917,0.361348,0.371508,0.388343,0.4142611,0.21863,0.273358,0.352738,0.2309952,0.379232,0.441279,0.299279,0.193936,0.4166806,0.313134,0.0,0.3232528,0.441934,0.430678,0.0,0.431892,0.42614,0.371262,0.415731,0.350579,0.393496,0.331932,0.409575,0.0,0.355582,0.413848,0.224602,0.379906,0.40699,"""S01""",7,"""WD1"""
"""M01D01""",2024,1,1,2024-01-01 00:00:00,"""NCAP_AF""","""ELCWNO""",0.193936,0.262934,0.193936,0.443864,0.359826,0.41041,0.43845,0.4064656,0.444845,0.43347,0.338858,0.2309952,0.379232,0.355065,0.45181,0.431095,0.406374,0.4370968,0.0,0.297433,0.415498,0.429774,0.45518,0.283774,0.396456,0.427897,0.382378,0.305921,0.426041,0.193936,0.437992,0.193936,0.3762425,0.392115,0.265779,0.4342062,0.242139,"""S01""",1,"""WD1"""
"""M01D01""",2025,1,1,2025-01-01 00:00:00,"""NCAP_AF""","""ELCWNO""",0.193936,0.3299185,0.278772,0.414718,0.416133,0.292333,0.429802,0.449838,0.440294,0.429321,0.0,0.4405172,0.379232,0.443026,0.447819,0.4037,0.4316286,0.406389,0.193936,0.317467,0.396437,0.369697,0.436877,0.438531,0.45555,0.435728,0.43697,0.431896,0.44654,0.331932,0.424215,0.30789,0.414574,0.434493,0.440889,0.4513218,0.350728,"""S01""",3,"""WD2"""
"""M01D01""",2026,1,1,2026-01-01 00:00:00,"""NCAP_AF""","""ELCWNO""",0.0,0.0,0.392696,0.319486,0.33711,0.343748,0.438829,0.447945,0.432704,0.352617,0.351088,0.3499146,0.0,0.4461544,0.435025,0.374674,0.292505,0.3117348,0.0,0.0,0.3221,0.366664,0.193936,0.283902,0.437596,0.412258,0.237885,0.392526,0.37333,0.436877,0.382796,0.331932,0.0,0.212624,0.375607,0.436095,0.409213,"""S01""",4,"""WD2"""
"""M01D01""",2027,1,1,2027-01-01 00:00:00,"""NCAP_AF""","""ELCWNO""",0.193936,0.397935,0.30673,0.441401,0.422829,0.193936,0.216935,0.193936,0.439896,0.445129,0.193936,0.193936,0.0,0.352173,0.447644,0.420702,0.444993,0.440904,0.0,0.0,0.412702,0.286584,0.438555,0.399942,0.0,0.43762,0.0,0.205436,0.416586,0.193936,0.389343,0.0,0.0,0.262934,0.262934,0.2447948,0.419606,"""S01""",5,"""WD2"""
"""M01D01""",2028,1,1,2028-01-01 00:00:00,"""NCAP_AF""","""ELCWNO""",0.0,0.317467,0.360714,0.425916,0.378818,0.446034,0.454664,0.4509236,0.445

# TS - 8760
Daily/hourly values. Ex: January 1, 2020 in USA

In [56]:

SSP126_TIMES.head()

TimeSlice,Yr,Month,Day,time,Attribute,PSET_CI,Virginia,Rhode Island,Season,Weekday,Type
str,i64,i64,i64,datetime[μs],str,str,f64,f64,str,i64,str
"""M01D01""",2020,1,1,2020-01-01 00:00:00,"""NCAP_AF""","""ELCWNF""",0.4663,0.379232,"""S01""",3,"""WD2"""
"""M01D01""",2021,1,1,2021-01-01 00:00:00,"""NCAP_AF""","""ELCWNF""",0.470652,0.436877,"""S01""",5,"""WD2"""
"""M01D01""",2022,1,1,2022-01-01 00:00:00,"""NCAP_AF""","""ELCWNF""",0.314227,0.193936,"""S01""",6,"""WD2"""
"""M01D01""",2023,1,1,2023-01-01 00:00:00,"""NCAP_AF""","""ELCWNF""",0.468683,0.436877,"""S01""",7,"""WD1"""
"""M01D01""",2024,1,1,2024-01-01 00:00:00,"""NCAP_AF""","""ELCWNF""",0.109932,0.421845,"""S01""",1,"""WD1"""


In [126]:
# save to csv 
SSP126_TIMES.write_csv('SSP126_Onshore_TS8760_TUSM.csv') # Onshore
#SSP126_TIMES.write_csv('SSP126_Offshore_TS8760_TUSM.csv') # Offshore

# TS - 120
Averaged by month and weekday type by country. ex: Weekdays in January in 2020 in USA, Weekends in January in 2020 in USA

In [127]:
# Group by year, month, type, then calculate the average for each groupby 
#groupby the Yr and Season and find mean 
SSP126_TIMES_120 = SSP126_TIMES.groupby(['Yr', 'Month', 'Type']).mean()
# sort by year and season
SSP126_TIMES_120 = SSP126_TIMES_120.sort(['Yr', 'Month'])
# make all rows in the 'PSET_CI' column 'ELCWNF' in polars
# add a new column called 'PSET_CI' that is 'ELCWNO' for onshore or 'ELCWNF' for offshore
SSP126_TIMES_120 = SSP126_TIMES_120.with_columns(pl.col('PSET_CI').fill_null('ELCWNF'))
# do the same but RNW for Attribute
SSP126_TIMES_120 = SSP126_TIMES_120.with_columns(pl.col('Attribute').fill_null('NCAP_AF'))
# make a column called 'M' that is just 'M' + Month
SSP126_TIMES_120 = SSP126_TIMES_120.with_columns(pl.lit('M').alias('M'))
#create a new column called 'TimeSlice' that is 'M' + Month + Type
SSP126_TIMES_120 = SSP126_TIMES_120.with_columns(pl.concat_str([pl.col('M'), pl.col('Month').cast(pl.Utf8).str.zfill(2), pl.col('Type')], separator='').alias('TimeSlice'))
# drop TimeSlice, time, Attribute, and PSET_CI columns
SSP126_TIMES_120 = SSP126_TIMES_120.drop(['Day', 'time', 'Type', 'M', 'Season', 'Weekday', 'Month'])

SSP126_TIMES_120

Yr,TimeSlice,Attribute,PSET_CI,Alaska,Arizona,California,Colorado,Hawaii,Idaho,Illinois,Indiana,Iowa,Kansas,Maine,Maryland,Massachusetts,Michigan,Minnesota,Missouri,Montana,Nebraska,Nevada,New Hampshire,New Mexico,New York,North Carolina,North Dakota,Ohio,Oklahoma,Oregon,Pennsylvania,South Dakota,Tennessee,Texas,Utah,Vermont,Washington,West Virginia,Wisconsin,Wyoming
i64,str,str,str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
2020,"""M01WD2""","""NCAP_AF""","""ELCWNF""",0.288039,0.250159,0.255944,0.378521,0.389275,0.289034,0.409977,0.414795,0.403314,0.396107,0.370151,0.39777,0.329624,0.416217,0.396851,0.379282,0.36974,0.398963,0.084704,0.316896,0.414953,0.405982,0.340421,0.375449,0.418187,0.398384,0.262129,0.3968,0.380576,0.26348,0.403928,0.184004,0.333138,0.240074,0.40167,0.415271,0.395379
2020,"""M01WD1""","""NCAP_AF""","""ELCWNF""",0.294501,0.319813,0.269115,0.420025,0.392751,0.315685,0.429062,0.429998,0.424589,0.394176,0.405151,0.4137264,0.366522,0.422484,0.434864,0.404668,0.395088,0.42972,0.048484,0.366348,0.413639,0.429686,0.380988,0.41497,0.412205,0.37261,0.251684,0.430091,0.431592,0.2988865,0.399944,0.222003,0.371952,0.209207,0.410489,0.420618,0.3909
2020,"""M02WD2""","""NCAP_AF""","""ELCWNF""",0.205703,0.306761,0.265996,0.385812,0.384998,0.272913,0.385245,0.3876,0.407997,0.392377,0.356593,0.376401,0.293517,0.414512,0.399695,0.371483,0.370907,0.39456,0.027705,0.313654,0.414604,0.368387,0.322526,0.391141,0.395311,0.403125,0.224199,0.377431,0.390145,0.233981,0.411967,0.245942,0.306633,0.210539,0.380669,0.382917,0.404372
2020,"""M02WD1""","""NCAP_AF""","""ELCWNF""",0.1142175,0.236495,0.230173,0.407154,0.3908,0.263729,0.431726,0.426677,0.431833,0.422966,0.375084,0.3864399,0.396016,0.434963,0.418997,0.422529,0.404697,0.423555,0.0,0.376405,0.40831,0.420811,0.3463655,0.400437,0.414389,0.40385,0.210756,0.417887,0.401445,0.216781,0.381328,0.220352,0.375047,0.215463,0.347265,0.426494,0.400748
2020,"""M03WD1""","""NCAP_AF""","""ELCWNF""",0.1937415,0.319927,0.204592,0.390883,0.393823,0.283513,0.400938,0.400984,0.432186,0.426088,0.397068,0.354746,0.3369416,0.413707,0.426507,0.425481,0.394066,0.424058,0.09524,0.358171,0.425636,0.392203,0.356888,0.39535,0.36948,0.426757,0.2235,0.376583,0.414658,0.293854,0.433626,0.2399564,0.35677,0.234986,0.343701,0.413346,0.386649
2020,"""M03WD2""","""NCAP_AF""","""ELCWNF""",0.150675,0.351469,0.219432,0.412418,0.379119,0.274589,0.406653,0.408387,0.423568,0.420994,0.352889,0.343531,0.330425,0.42668,0.428213,0.410776,0.383092,0.416152,0.08391,0.303956,0.421752,0.391134,0.353533,0.428642,0.412463,0.423524,0.203977,0.369815,0.425629,0.227099,0.419267,0.270041,0.338852,0.178921,0.347111,0.42568,0.410617
2020,"""M04WD2""","""NCAP_AF""","""ELCWNF""",0.151034,0.370975,0.269516,0.405332,0.381933,0.274895,0.37523,0.35076,0.428354,0.421221,0.338237,0.268167,0.270084,0.417271,0.431201,0.400394,0.360007,0.425062,0.189219,0.272315,0.415876,0.363802,0.377244,0.412109,0.32728,0.431079,0.310345,0.336053,0.426408,0.173842,0.43061,0.355165,0.297236,0.333765,0.291078,0.415095,0.409985
2020,"""M04WD1""","""NCAP_AF""","""ELCWNF""",0.157768,0.346537,0.288114,0.409995,0.369941,0.28896,0.415125,0.431668,0.397403,0.388834,0.342321,0.366009,0.235413,0.439431,0.417233,0.361442,0.374019,0.389879,0.072726,0.2301,0.399507,0.406084,0.3339995,0.400487,0.441324,0.422963,0.335759,0.379273,0.412878,0.336953,0.422677,0.283301,0.282344,0.339494,0.376748,0.435739,0.405605
2020,"""M05WD2""","""NCAP_AF""","""ELCWNF""",0.201894,0.321266,0.334478,0.377017,0.409133,0.325196,0.357674,0.350722,0.390689,0.393531,0.349265,0.198968,0.264292,0.417016,0.407055,0.362165,0.356416,0.394811,0.088581,0.273302,0.398888,0.338773,0.347092,0.405857,0.34124,0.386064,0.325672,0.293036,0.405998,0.159017,0.410348,0.240965,0.277627,0.321798,0.225178,0.371938,0.374012


In [128]:
# save to csv 
SSP126_TIMES_120.write_csv('SSP126_Onshore_TS120_TUSM.csv') # Onshore
#SSP126_TIMES_120.write_csv('SSP126_Offshore_TS120_TUSM.csv') # Offshore

# TS - 40
Averaged by season and weekday type by country. ex: Weekdays in Fall in 2020 in USA, Weekends in Fall in 2020 in USA

In [129]:
#groupby the Yr and Season and find mean 
SSP126_TIMES_40 = SSP126_TIMES.groupby(['Yr', 'Season', 'Type']).mean()
# sort by year and season
SSP126_TIMES_40 = SSP126_TIMES_40.sort(['Yr', 'Season'])
# drop TimeSlice, time, Attribute, and PSET_CI columns
SSP126_TIMES_40 = SSP126_TIMES_40.drop(['TimeSlice', 'time', 'Weekday', 'Month', 'Day'])
# make all rows in the 'PSET_CI' column 'ELCWNF' in polars
# add a new column called 'PSET_CI' that is 'ELCWNO' for onshore or 'ELCWNF' for offshore
SSP126_TIMES_40 = SSP126_TIMES_40.with_columns(pl.col('PSET_CI').fill_null('ELCWNF'))
# do the same but RNW for Attribute
SSP126_TIMES_40 = SSP126_TIMES_40.with_columns(pl.col('Attribute').fill_null('NCAP_AF'))
#Create a column called 'TimeSlice' that is the Season + Type
SSP126_TIMES_40 = SSP126_TIMES_40.with_columns(pl.concat_str([pl.col('Season'), pl.col('Type')], separator='').alias('TimeSlice'))
# drop the Season and Type columns
SSP126_TIMES_40 = SSP126_TIMES_40.drop(['Season', 'Type'])

SSP126_TIMES_40

Yr,Attribute,PSET_CI,Alaska,Arizona,California,Colorado,Hawaii,Idaho,Illinois,Indiana,Iowa,Kansas,Maine,Maryland,Massachusetts,Michigan,Minnesota,Missouri,Montana,Nebraska,Nevada,New Hampshire,New Mexico,New York,North Carolina,North Dakota,Ohio,Oklahoma,Oregon,Pennsylvania,South Dakota,Tennessee,Texas,Utah,Vermont,Washington,West Virginia,Wisconsin,Wyoming,TimeSlice
i64,str,str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,str
2020,"""NCAP_AF""","""ELCWNF""",0.182352,0.29403,0.266923,0.414427,0.381333,0.274983,0.402573,0.413101,0.405649,0.406904,0.360539,0.362642,0.360652,0.429346,0.408162,0.40379,0.397658,0.418079,0.024242,0.328356,0.401672,0.421524,0.316487,0.404791,0.41311,0.398287,0.233903,0.401981,0.401088,0.259408,0.39661,0.204669,0.361246,0.221384,0.361544,0.40742,0.401618,"""S01WD1"""
2020,"""NCAP_AF""","""ELCWNF""",0.236909,0.283904,0.262006,0.369859,0.39295,0.275451,0.39141,0.39048,0.402487,0.389165,0.361821,0.359634,0.317461,0.412205,0.385429,0.370529,0.370678,0.391181,0.054294,0.31916,0.404641,0.386605,0.342935,0.374101,0.393692,0.39434,0.239711,0.37759,0.374115,0.247803,0.405864,0.208825,0.322796,0.246597,0.364334,0.392486,0.393919,"""S01WD2"""
2020,"""NCAP_AF""","""ELCWNF""",0.21209,0.325725,0.26158,0.402339,0.388854,0.313665,0.394037,0.402379,0.402108,0.399268,0.353673,0.349419,0.293642,0.414247,0.407784,0.362593,0.376227,0.392206,0.084471,0.290358,0.404471,0.397044,0.340949,0.401396,0.396882,0.405195,0.297313,0.377228,0.40178,0.294722,0.419635,0.264797,0.312254,0.291729,0.357096,0.40407,0.391624,"""S02WD1"""
2020,"""NCAP_AF""","""ELCWNF""",0.168132,0.347848,0.275322,0.398038,0.39023,0.291821,0.37944,0.369365,0.41406,0.411776,0.346703,0.269094,0.287618,0.420224,0.422063,0.390809,0.36625,0.411945,0.121134,0.282872,0.412025,0.364161,0.359378,0.415334,0.359525,0.413402,0.281167,0.332401,0.419248,0.186031,0.420088,0.289011,0.304044,0.279688,0.286876,0.403908,0.398014,"""S02WD2"""
2020,"""NCAP_AF""","""ELCWNF""",0.134001,0.33058,0.325128,0.316682,0.391283,0.254012,0.310315,0.287142,0.347552,0.35238,0.276764,0.168724,0.192236,0.362244,0.355182,0.32204,0.316374,0.368029,0.06516,0.195714,0.355018,0.29715,0.211065,0.364333,0.271645,0.372149,0.310064,0.22449,0.363671,0.072463,0.396194,0.265502,0.204926,0.313316,0.152417,0.317902,0.316136,"""S03WD2"""
2020,"""NCAP_AF""","""ELCWNF""",0.124806,0.336882,0.343875,0.356029,0.397225,0.254246,0.288815,0.305296,0.305468,0.322947,0.279389,0.244233,0.200828,0.361678,0.344064,0.265266,0.327409,0.315449,0.101702,0.207305,0.36474,0.339597,0.361776,0.363119,0.287144,0.349028,0.324759,0.267722,0.340873,0.084151,0.384257,0.212551,0.230852,0.324183,0.205703,0.296117,0.313667,"""S03WD1"""
2020,"""NCAP_AF""","""ELCWNF""",0.190784,0.283891,0.237592,0.376927,0.383622,0.285409,0.349296,0.362015,0.391356,0.382668,0.359253,0.307194,0.309392,0.38999,0.395015,0.36531,0.358661,0.392624,0.083681,0.313836,0.398496,0.359291,0.337905,0.379296,0.352703,0.388019,0.263624,0.344766,0.383793,0.190545,0.394189,0.258704,0.325374,0.291942,0.314341,0.371382,0.391355,"""S04WD2"""
2020,"""NCAP_AF""","""ELCWNF""",0.186864,0.275752,0.285502,0.372519,0.386941,0.261018,0.366658,0.361682,0.399288,0.393586,0.344368,0.301772,0.265268,0.410444,0.395773,0.352728,0.362681,0.39002,0.09336,0.283423,0.396034,0.370731,0.323293,0.369119,0.351253,0.394596,0.264136,0.331308,0.373641,0.156992,0.391774,0.208047,0.305311,0.292245,0.339278,0.399544,0.394138,"""S04WD1"""
2021,"""NCAP_AF""","""ELCWNF""",0.211239,0.271426,0.285347,0.3899,0.393666,0.318114,0.380178,0.37958,0.399642,0.394179,0.366754,0.329183,0.2935826,0.396199,0.393255,0.366367,0.358325,0.392494,0.085535,0.305389,0.399011,0.370839,0.299319,0.379415,0.363628,0.396443,0.279301,0.359639,0.385683,0.222403,0.404653,0.276823,0.30244,0.296377,0.336693,0.396302,0.39965,"""S01WD2"""


In [131]:
# save to csv 
SSP126_TIMES_40.write_csv('SSP126_Onshore_TS40_TUSM.csv') # Onshore
#SSP126_TIMES_40.write_csv('SSP126_Offshore_TS40_TUSM.csv') # Offshore

# TS - 1
Averaged annually by country. ex: 2020 in USA

In [132]:

#groupby the Yr and find mean 
SSP126_TIMES_1 = SSP126_TIMES.groupby(['Yr']).mean()
# sort by year and season
SSP126_TIMES_1 = SSP126_TIMES_1.sort(['Yr'])
# drop TimeSlice, time, Attribute, and PSET_CI columns
SSP126_TIMES_1 = SSP126_TIMES_1.drop(['TimeSlice', 'time', 'Weekday', 'Month', 'Day', 'Season', 'Type'])
# make all rows in the 'PSET_CI' column 'ELCWNF' in polars
# add a new column called 'PSET_CI' that is 'ELCWNO' for onshore or 'ELCWNF' for offshore
SSP126_TIMES_1 = SSP126_TIMES_1.with_columns(pl.col('PSET_CI').fill_null('ELCWNF'))
# do the same but RNW for Attribute
SSP126_TIMES_1 = SSP126_TIMES_1.with_columns(pl.col('Attribute').fill_null('NCAP_AF'))
SSP126_TIMES_1.head()

Yr,Attribute,PSET_CI,Alaska,Arizona,California,Colorado,Hawaii,Idaho,Illinois,Indiana,Iowa,Kansas,Maine,Maryland,Massachusetts,Michigan,Minnesota,Missouri,Montana,Nebraska,Nevada,New Hampshire,New Mexico,New York,North Carolina,North Dakota,Ohio,Oklahoma,Oregon,Pennsylvania,South Dakota,Tennessee,Texas,Utah,Vermont,Washington,West Virginia,Wisconsin,Wyoming
i64,str,str,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
2020,"""NCAP_AF""","""ELCWNF""",0.180993,0.310626,0.279245,0.371162,0.389339,0.276518,0.359002,0.357354,0.385604,0.382831,0.335568,0.287157,0.277252,0.398233,0.389089,0.357158,0.356532,0.387183,0.079841,0.277534,0.392286,0.360346,0.319656,0.383473,0.349303,0.390397,0.275692,0.326725,0.383318,0.181222,0.402413,0.246066,0.292709,0.283018,0.289865,0.372747,0.374808
2021,"""NCAP_AF""","""ELCWNF""",0.204236,0.318308,0.290339,0.392922,0.393359,0.309453,0.350734,0.343105,0.389896,0.40217,0.31352,0.257674,0.240602,0.393128,0.391205,0.363367,0.358462,0.397045,0.104939,0.247249,0.406496,0.334899,0.298723,0.390457,0.328402,0.403147,0.285707,0.30155,0.393714,0.160979,0.412459,0.280445,0.265412,0.298011,0.268369,0.366676,0.388531
2022,"""NCAP_AF""","""ELCWNF""",0.187651,0.317416,0.278891,0.393512,0.383084,0.312276,0.362137,0.349029,0.394505,0.397881,0.335013,0.254557,0.248946,0.392781,0.395823,0.365603,0.36377,0.397809,0.099763,0.262108,0.405087,0.342247,0.31194,0.391346,0.332518,0.398694,0.304847,0.303651,0.397158,0.169426,0.41058,0.262773,0.281178,0.308552,0.261587,0.371924,0.396001
2023,"""NCAP_AF""","""ELCWNF""",0.194789,0.335154,0.280802,0.391081,0.383281,0.309683,0.335745,0.334538,0.380709,0.400256,0.313028,0.241161,0.221662,0.391063,0.384302,0.362137,0.368524,0.397179,0.126165,0.232179,0.406231,0.331056,0.308264,0.397743,0.321628,0.400927,0.294428,0.293676,0.397045,0.171852,0.417079,0.276178,0.253515,0.300427,0.251464,0.35675,0.395413
2024,"""NCAP_AF""","""ELCWNF""",0.179206,0.323032,0.270077,0.392277,0.385987,0.317934,0.355573,0.348234,0.384491,0.399132,0.301427,0.242926,0.210011,0.38994,0.390486,0.364857,0.369158,0.395746,0.116712,0.219555,0.402566,0.323839,0.29357,0.400419,0.328695,0.399719,0.302672,0.287767,0.401846,0.153599,0.412625,0.271209,0.242501,0.302084,0.250924,0.352601,0.395021


In [133]:
# save to csv 
SSP126_TIMES_1.write_csv('SSP126_Onshore_TS1_TUSM.csv') # Onshore
#SSP126_TIMES_1.write_csv('SSP126_Offshore_TS1_TUSM.csv') # Offshore